akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -218,3 +225,191 @@ private[fusing] object StreamOfStreams {
     }
   }
 }
+
+/**
+ * INTERNAL API
+ */
+object PrefixAndTail {
+
+  sealed trait MaterializationState
+  case object NotMaterialized extends MaterializationState
+  case object AlreadyMaterialized extends MaterializationState
+  case object TimedOut extends MaterializationState
+
+  case object NormalCompletion extends MaterializationState
+  case class FailureCompletion(ex: Throwable) extends MaterializationState
+
+  trait TailInterface[T] {
+    def pushSubstream(elem: T): Unit
+    def completeSubstream(): Unit
+    def failSubstream(ex: Throwable)
+  }
+
+  final class TailSource[T](
+    timeout: FiniteDuration,
+    register: TailInterface[T] ⇒ Unit,
+    pullParent: Unit ⇒ Unit,
+    cancelParent: Unit ⇒ Unit) extends GraphStage[SourceShape[T]] {
+    val out: Outlet[T] = Outlet("Tail.out")
+    val materializationState = new AtomicReference[MaterializationState](NotMaterialized)
+    override val shape: SourceShape[T] = SourceShape(out)
+
+    private final class TailSourceLogic(_shape: Shape) extends GraphStageLogic(_shape) with OutHandler with TailInterface[T] {
+      setHandler(out, this)
+
+      override def preStart(): Unit = {
+        materializationState.getAndSet(AlreadyMaterialized) match {
+          case AlreadyMaterialized ⇒
+            failStage(new IllegalStateException("Tail Source cannot be materialized more than once."))
+          case TimedOut ⇒
+            // Already detached from parent
+            failStage(new TimeoutException(s"Tail Source has not been subscribed in $timeout."))
```

Comment:
That might be only true if I accidentally (incidentally?) fixed that issue (and now there is only an Exception mismatch). See other PR validation (https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/4552/consoleFull): 

```
[info] SubstreamSubscriptionTimeoutSpec:
Coroner Thread Count starts at 25 in akka.stream.scaladsl.SubstreamSubscriptionTimeoutSpec
[info] groupBy
[info] - must timeout and cancel substream publishers when no-one subscribes to them after some time (time them out) *** FAILED *** (1 second, 76 milliseconds)
[info]   3 did not equal "expected" (SubstreamSubscriptionTimeoutSpec.scala:80)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.MatchersHelper$.newTestFailedException(MatchersHelper.scala:160)
[info]   at org.scalatest.Matchers$ShouldMethodHelper$.shouldMatcher(Matchers.scala:6231)
[info]   at org.scalatest.Matchers$AnyShouldWrapper.should(Matchers.scala:6277)
[info]   at akka.stream.scaladsl.SubstreamSubscriptionTimeoutSpec$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$2.apply$mcV$sp(SubstreamSubscriptionTimeoutSpec.scala:80)
[info]   at akka.stream.scaladsl.SubstreamSubscriptionTimeoutSpec$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$2.apply(SubstreamSubscriptionTimeoutSpec.scala:43)
[info]   at 
```

That test fails on 2.3-dev and on my machine as well, without this patch.